### PR TITLE
Remove assert, use typeforce typing

### PR DESCRIPTION
**Pull request by dcousens**
Will wait on ACK before merge.

ping @weilu, @jprichardson 

Curious on your opinions about this...

### CODE REVIEW

File: src/types.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  return typeforce.Number(value) &&
+    value >= 0 &&
+    value <= UINT53_MAX &&
+    Math.floor(value) === value
+}
+
+// external dependent types
+function BigInt (value) { return !typeforce.Null(value) && value.constructor === bigi }
+function ECCurve (value) { return !typeforce.Null(value) && value.constructor === ecurve.Curve }
+function ECPoint (value) { return !typeforce.Null(value) && value.constructor === ecurve.Point }
```
Comment (dcousens) [inline src/types.js]: @weilu you this does revert our duck typing changes.   I've never had that problem,  so I'm not sure why we allowed it in the first place.

`ECPair` and `ecdsa` are the only interfaces that accept `bigi` / `ECPoint`'s,  and we [intend](https://github.com/bitcoinjs/bitcoinjs-lib/issues/407) to [remove](https://github.com/bitcoinjs/bitcoinjs-lib/issues/394) their individual use entirely for `2.0.0`,  so this shouldn't even be an issue.


Various bugs where internal versions of these libraries could be inconsistent means duck typing is a very bad idea in this case.

File: test/fixtures/transaction.json [hunk 2]
```diff
@@ -229,12 +229,12 @@
   "invalid": {
     "addInput": [
       {
-        "exception": "Expected hash length of 32, got 30",
+        "exception": "Expected 256-bit Buffer, got 240-bit",
```
Comment (dcousens) [inline test/fixtures/transaction.json]: Should this error be `Expected 256-bit hash, got 240-bit`?

File: src/transaction.js [hunk 3]
```diff
@@ -19,7 +19,7 @@ Transaction.SIGHASH_NONE = 0x02
 Transaction.SIGHASH_SINGLE = 0x03
 Transaction.SIGHASH_ANYONECANPAY = 0x80
 
-Transaction.fromBuffer = function (buffer, __disableAssert) {
+Transaction.fromBuffer = function (buffer, __disableExcess) {
```
Comment (dcousens) [inline src/transaction.js]: `__disableExcess`, if provides, disables the 'strictness' of `fromBuffer` from checking that the bytes read are equal to the bytes in `buffer`.

File: src/ecdsa.js [hunk 4]
```diff
@@ -10,12 +10,12 @@ var ONE = new Buffer([1])
...
-  typeForce('Function', checkSig)
-
-  // sanity check
-  assert.equal(hash.length, 32, 'Hash must be 256 bit')
+  typeforce(types.tuple(
+    types.ECCurve,
+    types.Hash256bit,
+    types.BigInt,
+    types.Function
+  ), arguments)
```
Comment (dcousens) [inline src/ecdsa.js]: All these types are currently being created on the fly (though, it is pretty light weight), we can pull them out of the function and just enforce them on the first line if we feel performance is an issue,  though,  this new syntax is already 1000% faster than the previous implementation.

**Comment by dcousens** [comment]
I'm happy to break out the `typeforce` typing changes if theres any issues,  as the move away from `assert` is a much simpler decision and it does have highly tangible benefits.

**Comment by jprichardson** [comment]
> as the move away from assert is a much simpler decision and it does have highly tangible benefits.

I think for posterity, these benefits should be explicitly stated.

File: src/types.js (hunk 1 above)
Comment (weilu) [inline src/types.js]: I remember an important reason why we went with duck typing is to deal with having multiple versions of the same library in the same app. I still think that's relevant so my preference is still duck typing.

File: src/transaction.js (hunk 3 above)
Comment (weilu) [inline src/transaction.js]: `__lenient` or `__noCheckTransactionData`. I find negated names most of the time confusing. Given that you prepended it with __ I wonder if we can get away with flipping it to `__strict` without breaking the API.

File: test/fixtures/transaction.json (hunk 2 above)
Comment (weilu) [inline test/fixtures/transaction.json]: I think Buffer is fine

**Comment by weilu** [comment]
I don't get why you are getting rid of asserts. IMHO in most of the places, assert is still more direct and semantically straightforward. It's not like we are throwing any custom Error types for better error handling.

**Comment by dcousens** [comment]
> I think for posterity, these benefits should be explicitly stated.

* In a production application, with over 30 high level dependencies,  `bitcoinjs-lib`, `bs58`, `bip39`, `bs58check`,  are the only dependencies using `assert` for error handling.   All other conclusions aside,  this is a specific pattern we have adopted for our ecosystem.

* `assert` explicitly makes code coverage difficult,  because the branch is deeper down in the application.   `typeforce` changes do not help this,  however,  the majority of `typeforce` changes were `type` checking.  The majority of logic based changes that now exist as `throw` conditions are now visible in code coverage.

* `assert` only accepts value arguments (like any function), therefore error messages are evaluated/built on every call.   This has caused major performance problems for us in the past where `JSON.stringify` or `toString` has been more complex then expected and excessively slowed down applications.

* `assert.equal` is more often used over `assert.strictEqual`,  most likely because it is shorter to write.  We actually have our linter fail our tests if we wrote this code normally, and yet, we have no qualms with this.  Granted,  that could just be fixed by enforcing `strictEqual`,  it is more verbose and often longer code than just writing out the explicit branch anyway.

File: src/types.js (hunk 1 above)
Comment (dcousens) [inline src/types.js]: > I remember an important reason why we went with duck typing is to deal with having multiple versions of the same library in the same app.

Right.

> I've never had the issue of de-duping where it wasn't symptomatic of something potentially disastrous anyway, so I'm not sure why we allowed it in the first place.

I feel like such a situation is dangerous enough that we shouldn't let it occur though?

File: src/transaction.js (hunk 3 above)
Comment (dcousens) [inline src/transaction.js]: @weilu the default is strict,  it is only in [specific cases](https://github.com/bitcoinjs/bitcoinjs-lib/blob/master/src/block.js#L47-L53) that you would want to do this.

`__noStrict` or `__lenient` works for me.

**Comment by dcousens** [comment]
Rebased on master.  Added 94e16fb,  6002d9a.

**Comment by dcousens** [comment]
@jprichardson @weilu,  shall I merge or?

**Comment by weilu** [comment]
@dcousens Given that the types of the errors and in some cases the messages are different, I wonder if we need to bump major for this?

**Comment by dcousens** [comment]
@weilu I wasn't going to bother back porting this to `1.5.x`,  the change is worth it,  but it doesn't fix/add anything such that it would be worth back porting.

File: test/fixtures/transaction_builder.json [hunk 5]
```diff
@@ -780,7 +780,7 @@
       },
       {
         "description": "Wrong key pair for multisig redeemScript",
-        "exception": "key pair cannot sign for this input",
+        "exception": "Key pair cannot sign for this input",
```
Comment (dcousens) [inline test/fixtures/transaction_builder.json]: Breaking change

File: test/fixtures/transaction.json [hunk 6]
```diff
@@ -229,12 +229,12 @@
...
         "hash": "0aed1366a73b6057ee7800d737bff1bdf8c448e98d86bc0998f2b009816d",
         "index": 0
       },
       {
-        "exception": "Expected hash length of 32, got 34",
+        "exception": "Expected 256-bit Buffer, got 272-bit",
```
Comment (dcousens) [inline test/fixtures/transaction.json]: Breaking change

File: test/fixtures/transaction.json (hunk 2 above)
Comment (dcousens) [inline test/fixtures/transaction.json]: Breaking change

File: test/fixtures/ecdsa.json [hunk 7]
```diff
@@ -188,7 +188,7 @@
       },
       {
         "description": "Invalid i value (> 3)",
-        "exception": "Recovery param is more than two bits",
+        "exception": "Expected UInt2, got Number 4",
```
Comment (dcousens) [inline test/fixtures/ecdsa.json]: Breaking change

**Comment by dcousens** [comment]
I'd say there is only a few breaking changes, outlined above.  And even then,  `ecdsa` is no longer exposed.   The error messages that are different are therefore only for `Transaction`,  the rest should be consistent as `types.tuple` should behave the same as the previous code.

File: src/address.js [hunk 8]
```diff
@@ -23,10 +24,7 @@ function fromOutputScript (script, network) {
 }
 
 function toBase58Check (hash, version) {
-  typeForce('Buffer', hash)
-
-  if (hash.length !== 20) throw new TypeError('Invalid hash length')
-  if (version & ~0xff) throw new TypeError('Invalid version byte')
+  typeforce(types.tuple(types.Hash160bit, types.UInt8), arguments)
```
Comment (dcousens) [inline src/address.js]: This API is new,  so its not a breaking change.

File: src/types.js (hunk 1 above)
Comment (dcousens) [inline src/types.js]: This is the other breaking change,  but,  that has already been pointed out.

**Comment by weilu** [comment]
I'm fine with not back-porting this to 1.5.x but let's add these breaking changes into our 2.0 release note.

**Comment by weilu** [comment]
Good to merge

**Comment by dcousens** [comment]
Thanks @weilu,  I added the breaking change tag so we can easily build that release note by reviewing it later.

**Comment by dcousens** [comment]
Follow up https://github.com/bitcoinjs/bitcoinjs-lib/commit/5d2abb523ee32c7b9039330152134825c0a7c918 for `HDNode`.

File: src/types.js (hunk 1 above)
Comment (dcousens) [inline src/types.js]: This change was reverted in https://github.com/bitcoinjs/bitcoinjs-lib/pull/450 as it was still an issue.  `quacksLike` behaviour was restored.

**Pull request merged**